consensus/istanbul/core/handler_test.go

```diff
@@ -464,6 +483,236 @@ func TestCore_handlerMsg(t *testing.T) {
 	}
 }
 
+// TODO-Klaytn: To enable logging in the test code, we can use the following function.
+// This function will be moved to somewhere utility functions are located.
+func enableLog() {
+	usecolor := term.IsTty(os.Stderr.Fd()) && os.Getenv("TERM") != "dumb"
+	output := io.Writer(os.Stderr)
+	if usecolor {
+		output = colorable.NewColorableStderr()
+	}
+	glogger := log.NewGlogHandler(log.StreamHandler(output, log.TerminalFormat(usecolor)))
+	log.PrintOrigins(true)
+	log.ChangeGlobalLogLevel(glogger, log.Lvl(3))
+	glogger.Vmodule("")
+	glogger.BacktraceAt("")
+	log.Root().SetHandler(glogger)
+}
+
+// splitSubList splits a committee into two groups w/o proposer
+// one for n nodes, the other for len(committee) - n - 1 nodes
+func splitSubList(committee []istanbul.Validator, n int, proposerAddr common.Address) ([]istanbul.Validator, []istanbul.Validator) {
+	var subCN, remainingCN []istanbul.Validator
+
+	for _, val := range committee {
+		if val.Address() == proposerAddr {
+			// proposer is not included in any group
+			continue
+		}
+		if len(subCN) < n {
+			subCN = append(subCN, val)
+		} else {
+			remainingCN = append(remainingCN, val)
+		}
+	}
+	return subCN, remainingCN
+}
+
+// Simulate a proposer that receives messages from disagreeing groups of CNs.
+func simulateMaliciousCN(t *testing.T, numValidators int, numMalicious int) State {
+	if testing.Verbose() {
+		enableLog()
+	}
+
+	fork.SetHardForkBlockNumberConfig(&params.ChainConfig{})
+	defer fork.ClearHardForkBlockNumberConfig()
+
+	// Note that genValidators(n) will generate n/3 validators.
+	// We want n validators, thus calling genValidators(3n).
+	validatorAddrs, validatorKeyMap := genValidators(numValidators * 3)
+
+	// Add more EXPECT()s to remove unexpected call error
+	mockBackend, mockCtrl := newMockBackend(t, validatorAddrs)
+	mockBackend.EXPECT().Commit(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	mockBackend.EXPECT().HasBadProposal(gomock.Any()).Return(true).AnyTimes()
+	defer mockCtrl.Finish()
+
+	var (
+		// it creates two pre-defined blocks: one for benign CNs, the other for the malicious
+		// newProposal is a block which the proposer has created
+		// malProposal is an incorrect block that malicious CNs use to try stop consensus
+		lastProposal, _ = mockBackend.LastProposal()
+		lastBlock       = lastProposal.(*types.Block)
+		validators      = mockBackend.Validators(lastBlock)
+		proposer        = validators.GetProposer()
+		proposerKey     = validatorKeyMap[proposer.Address()]
+		newProposal, _  = genBlockParams(lastBlock, proposerKey, 0, 1, 1)
+		malProposal, _  = genBlockParams(lastBlock, proposerKey, 0, 0, 0)
```

Comment:
What does distinguish good/bad proposal?

Reply:
A proposer generates a block as newProposal.
Malicious CNs do not accept the proposer's block and use malProposal's hash value to send consensus message.
I added a description of newProposal and malProposal as comment on the code.